indico/web/client/js/react/components/WTFOccurrencesField.jsx

```diff
@@ -0,0 +1,177 @@
+}
+
+export default function WTFOccurrencesField({fieldId, uses24HourFormat}) {
+  const field = useMemo(() => document.getElementById(fieldId), [fieldId]);
+  const [values, setValues] = useState(() =>
+    JSON.parse(field.value).map(value => ({
+      date: moment(value.date, 'YYYY-MM-DD'),
+      duration: value.duration,
+      time: moment(value.time, 'HH:mm'),
+      id: _.uniqueId(),
+    }))
+  );
+
+  const updateOccurrences = newValues => {
+    setValues(newValues);
+    const storageValues = newValues.map(value => ({
+      date: value.date ? value.date.format('YYYY-MM-DD') : '',
+      duration: value.duration,
+      time: value.time ? value.time.format('HH:mm') : '',
+    }));
+    field.value = JSON.stringify(storageValues);
+  };
+
+  const addOccurrence = () => {
+    const lastElement = values[values.length - 1];
+    const newValues = values.concat([
+      {
+        date: lastElement.date ? moment(lastElement.date).add(1, 'day') : null,
+        duration: lastElement.duration,
+        time: lastElement.time,
+        id: _.uniqueId(),
+      },
+    ]);
+    updateOccurrences(newValues);
+    triggerChange(field);
+  };
+
+  const changeOccurrence = (index, date, duration, time) => {
+    const newValues = [...values];
+    newValues[index] = {date, duration, time, id: values[index].id};
+    updateOccurrences(newValues);
+    triggerChange(field);
+  };
+
+  const removeOccurrence = index => {
+    updateOccurrences(values.filter((__, i) => index !== i));
+    triggerChange(field);
+  };
+
+  return (
+    <>
+      {values.map((element, i) => (
+        <div className="occurrence" key={element.id}>
+          <Occurrence
+            occurrence={element}
+            uses24HourFormat={uses24HourFormat}
+            changeOccurrence={changeOccurrence.bind(null, i)}
+            removeOccurrence={values.length > 1 ? removeOccurrence.bind(null, i) : null}
+          />
+        </div>
+      ))}
+      <div className="add-occurrence" onClick={addOccurrence}>
+        {Translate.string('Add occurrence')}
+      </div>
+    </>
+  );
+}
+
+WTFOccurrencesField.propTypes = {
+  fieldId: PropTypes.string.isRequired,
+  uses24HourFormat: PropTypes.bool.isRequired,
+};
+
+function Occurrence({occurrence, uses24HourFormat, changeOccurrence, removeOccurrence}) {
+  const format = uses24HourFormat ? 'HH:mm' : 'hh:mm a';
+  const clearRef = useRef(null);
+  const timePickerRef = useRef(null);
+
+  useEffect(() => {
+    timePickerRef.current.picker.required = true;
+  }, []);
+
+  const updateDate = value => {
+    changeOccurrence(value, occurrence.duration, occurrence.time);
+  };
+
+  const updateTime = value => {
+    changeOccurrence(occurrence.date, occurrence.duration, value);
+  };
+
+  const updateDuration = event => {
+    changeOccurrence(occurrence.date, parseInt(event.target.value, 10), occurrence.time);
+  };
+
+  return (
+    <>
+      <SingleDatePicker
+        id=""
```

Comment:
`react-dates` documentation mentions that this prop is required but they use a `defaultProp` (with the value `"date"`) nonetheless (I also noticed that in another usage of this component we don't pass an `id` prop). So, I think you can omit this.

Reply:
Nope. We actually fought with this one on the previous PR because if you don't set it to an empty string, it also adds `name="date"`, and thus breaks the logic to disable the submit button when there are no changes because then you have an additional named field in the form which gets included in determining its values...

So in cases where you don't care about the id and name of the underlying input field, you can indeed omit it. But in the wtforms context where the actual html form contents (as seen by jquery etc) matter, this is the only workaroud :/